### PR TITLE
Parser records missing units 

### LINKED ISSUE #686: Parser to create messages when it can't read stuff?
While working on the `Importer` functionality I had first-hand experience at how tricky it can be to figure out errors and unexpected behaviour in an imported model situation.  The biggest area that I struggled with was when the `Parser` didn't have any issues, but wasn't reading the file in the way I thought it should have been.  I'd like to take advantage of the different levels of `Issue` to create warnings/hints/messages when the `Parser` encounters anything that can't be instantiated into a model.

### PR DESCRIPTION
**Pull request by kerimoyle**
Addresses #686

### CODE REVIEW

File: src/utilities.cpp [hunk 1]
```diff
@@ -693,4 +695,70 @@ std::string makeUniqueId(IdList &idList)
...
 }
 
+void linkComponentVariableUnits(const ComponentPtr &component, const LoggerPtr &logger)
+{
+    for (size_t index = 0; index < component->variableCount(); ++index) {
+        auto v = component->variable(index);
+        auto u = v->units();
+        if (u != nullptr) {
+            auto model = owningModel(u);
+            if (model == nullptr && !isStandardUnit(u)) {
```
Comment (kerimoyle) [inline src/utilities.cpp]: Should this raise an issue when `model != nullptr` ? At present it just skips over it?

File: src/parser.cpp [hunk 2]
```diff
@@ -391,7 +391,16 @@ void Parser::ParserImpl::loadModel(const ModelPtr &model, const std::string &inp
         loadConnection(model, connectionNode);
     }
 
-    model->linkUnits();
+    auto logger = Logger::create();
+    for (size_t index = 0; index < model->componentCount(); ++index) {
+        auto c = model->component(index);
+        traverseComponentTreeLinkingUnits(c, logger);
```
Comment (kerimoyle) [inline src/parser.cpp]: Is there a cleaner way to pass the private logger in the parser into an external function?

**Comment by kerimoyle** [comment]
I haven't written docstrings yet because I want to know whether this implementation is going to fly or not first ... Mainly whether there is (and I think that there must be) a better way to access the internal `Logger` within the `Parser`?  See [here](https://github.com/cellml/libcellml/pull/687/files#r468997421).

Thoughts @agarny @nickerso @hsorby ?

File: src/api/libcellml/logger.h [hunk 3]
```diff
@@ -38,6 +38,20 @@ class LIBCELLML_EXPORT Logger
...
     Logger &operator=(Logger rhs) = delete; /**< Assignment operator */
 
+    /**
+     * @brief Create a @c Logger object.
+     *
+     * Factory method to create a @c Logger.  Create a
+     * blank logger with::
+     *
+     * @code
+     *   auto logger = libcellml::LOgger::create();
```
Comment (kerimoyle) [inline src/api/libcellml/logger.h]: Logger not LOgger

File: src/parser.cpp (hunk 2 above)
Comment (kerimoyle) [reply src/parser.cpp]: Pass `Parser` shared_from_this() instead of logger.

**Comment by kerimoyle** [comment]
Changes made following our discussion last night.  This is ready for review @hsorby @agarny @nickerso.

**Comment by kerimoyle** [comment]
I updated this branch last night, but it's still waiting to be built?  Any ideas what's going on @hsorby ?

**Review by nickerso — DISMISSED** [review]

**Comment by agarny** [comment]
@nickerso: this is ready for you to reapprove, if you still want to. :)

**Review by agarny — DISMISSED** [review]

**Review by nickerso — DISMISSED** [review]
not sure why we need our own string replace method in the utilities, but suspect github is showing me a change that was made in a different PR.

**Review by agarny — APPROVED** [review]

**Review by nickerso — APPROVED** [review]

**Pull request merged**